In [1]:
import pandas as pd
import numpy as np
import pickle

print(pd.__version__) # 1.1.3
print(np.__version__) # 1.19.2

1.1.3
1.19.2


# Importation des données

In [2]:
# import du data set customers et premières informations

raw_customers = pd.read_csv("archive/olist_customers_dataset.csv")
print(raw_customers.shape)
raw_customers.head()

(99441, 5)


customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [3]:
# statistiques sur raw_customers

print(raw_customers.dtypes)
raw_customers.describe(include="all")

customer_id                 object
customer_unique_id          object
customer_zip_code_prefix     int64
customer_city               object
customer_state              object
dtype: object


customer_id                customer_unique_id  \
count                              99441                             99441   
unique                             99441                             96096   
top     f81a949b0915fdaf3b6592c5e037d637  8d50f5eadf50201ccdcedfb9e2ac8455   
freq                                   1                                17   
mean                                 NaN                               NaN   
std                                  NaN                               NaN   
min                                  NaN                               NaN   
25%                                  NaN                               NaN   
50%                                  NaN                               NaN   
75%                                  NaN                               NaN   
max                                  NaN                               NaN   

        customer_zip_code_prefix customer_city customer_state  
count               99441.000000         99441          99441  
unique                       NaN          4119             27  
top                          NaN     sao paulo             SP  
freq                         NaN         15540          41746  
mean                35137.474583           NaN            NaN  
std                 29797.938996           NaN            NaN  
min                  1003.000000           NaN            NaN  
25%                 11347.000000           NaN            NaN  
50%                 24416.000000           NaN            NaN  
75%                 58900.000000           NaN            NaN  
max                 99990.000000           NaN            NaN

Il y a donc 96 096 clients différents pour 99 441 commandes, ils sont répartis dans 4119 villes et 27 états.  
Il n'y a pas de valeurs manquantes.

In [4]:
# import du data set items et premières informations

items = pd.read_csv("archive/olist_order_items_dataset.csv", parse_dates=["shipping_limit_date"])
print(items.shape)
items.head()

(112650, 7)


order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

  shipping_limit_date   price  freight_value  
0 2017-09-19 09:45:35   58.90          13.29  
1 2017-05-03 11:05:13  239.90          19.93  
2 2018-01-18 14:48:30  199.00          17.87  
3 2018-08-15 10:10:18   12.99          12.79  
4 2017-02-13 13:57:51  199.90          18.14

In [5]:
# statistiques sur items

print(items.dtypes)
items.describe(include="all")

order_id                       object
order_item_id                   int64
product_id                     object
seller_id                      object
shipping_limit_date    datetime64[ns]
price                         float64
freight_value                 float64
dtype: object


<ipython-input-5-50524efa58cd>:4: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  items.describe(include="all")


order_id  order_item_id  \
count                             112650  112650.000000   
unique                             98666            NaN   
top     8272b63d03f5f79c56e9e4120aec44ef            NaN   
freq                                  21            NaN   
first                                NaN            NaN   
last                                 NaN            NaN   
mean                                 NaN       1.197834   
std                                  NaN       0.705124   
min                                  NaN       1.000000   
25%                                  NaN       1.000000   
50%                                  NaN       1.000000   
75%                                  NaN       1.000000   
max                                  NaN      21.000000   

                              product_id                         seller_id  \
count                             112650                            112650   
unique                             32951                              3095   
top     aca2eb7d00ea1a7b8ebd4e68314663af  6560211a19b47992c3666cc44a7e94c0   
freq                                 527                              2033   
first                                NaN                               NaN   
last                                 NaN                               NaN   
mean                                 NaN                               NaN   
std                                  NaN                               NaN   
min                                  NaN                               NaN   
25%                                  NaN                               NaN   
50%                                  NaN                               NaN   
75%                                  NaN                               NaN   
max                                  NaN                               NaN   

        shipping_limit_date          price  freight_value  
count                112650  112650.000000  112650.000000  
unique                93318            NaN            NaN  
top     2018-03-01 02:50:48            NaN            NaN  
freq                     21            NaN            NaN  
first   2016-09-19 00:15:34            NaN            NaN  
last    2020-04-09 22:35:08            NaN            NaN  
mean                    NaN     120.653739      19.990320  
std                     NaN     183.633928      15.806405  
min                     NaN       0.850000       0.000000  
25%                     NaN      39.900000      13.080000  
50%                     NaN      74.990000      16.260000  
75%                     NaN     134.900000      21.150000  
max                     NaN    6735.000000     409.680000

Pas de valeurs manquantes, 32 951 produits répartis sur 98 666 commandes différentes.  
order_idem_id différencie les différents articles d'une même commande (que ce soit les mêmes en plusieurs exemplaires ou bien des différents).

In [6]:
# import du data set payments et premières informations

payments = pd.read_csv("archive/olist_order_payments_dataset.csv")
print(payments.shape)
payments.head()

(103886, 5)


order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

In [7]:
# statistiques sur payments

payments.describe(include="all")

order_id  payment_sequential payment_type  \
count                             103886       103886.000000       103886   
unique                             99440                 NaN            5   
top     fa65dad1b0e818e3ccc5cb0e39231352                 NaN  credit_card   
freq                                  29                 NaN        76795   
mean                                 NaN            1.092679          NaN   
std                                  NaN            0.706584          NaN   
min                                  NaN            1.000000          NaN   
25%                                  NaN            1.000000          NaN   
50%                                  NaN            1.000000          NaN   
75%                                  NaN            1.000000          NaN   
max                                  NaN           29.000000          NaN   

        payment_installments  payment_value  
count          103886.000000  103886.000000  
unique                   NaN            NaN  
top                      NaN            NaN  
freq                     NaN            NaN  
mean                2.853349     154.100380  
std                 2.687051     217.494064  
min                 0.000000       0.000000  
25%                 1.000000      56.790000  
50%                 1.000000     100.000000  
75%                 4.000000     171.837500  
max                24.000000   13664.080000

Pas de valeurs manquantes.  
99 940 commandes dont certaines réglées de différentes façons (sequential).

In [8]:
# import du data set reviews et premières informations

reviews = pd.read_csv("archive/olist_order_reviews_dataset.csv",
                      parse_dates = ["review_creation_date", "review_answer_timestamp"])
print(reviews.shape)
reviews.head()

(100000, 7)


review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN           2018-01-18   
1                                                NaN           2018-03-10   
2                                                NaN           2018-02-17   
3              Recebi bem antes do prazo estipulado.           2017-04-21   
4  Parabéns lojas lannister adorei comprar pela I...           2018-03-01   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

In [9]:
# statistiques sur reviews

print(reviews.dtypes)
reviews.describe(include="all")

review_id                          object
order_id                           object
review_score                        int64
review_comment_title               object
review_comment_message             object
review_creation_date       datetime64[ns]
review_answer_timestamp    datetime64[ns]
dtype: object


<ipython-input-9-f83feec60c6f>:4: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  reviews.describe(include="all")
<ipython-input-9-f83feec60c6f>:4: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  reviews.describe(include="all")


review_id                          order_id  \
count                             100000                            100000   
unique                             99173                             99441   
top     9e25d6e3025e9b9a0fc7f03588d33e2b  03c939fd7fd3b38f8485a0f95798f1f6   
freq                                   3                                 3   
first                                NaN                               NaN   
last                                 NaN                               NaN   
mean                                 NaN                               NaN   
std                                  NaN                               NaN   
min                                  NaN                               NaN   
25%                                  NaN                               NaN   
50%                                  NaN                               NaN   
75%                                  NaN                               NaN   
max                                  NaN                               NaN   

         review_score review_comment_title review_comment_message  \
count   100000.000000                11715                  41753   
unique            NaN                 4600                  36921   
top               NaN            Recomendo              Muito bom   
freq              NaN                  426                    230   
first             NaN                  NaN                    NaN   
last              NaN                  NaN                    NaN   
mean         4.070890                  NaN                    NaN   
std          1.359663                  NaN                    NaN   
min          1.000000                  NaN                    NaN   
25%          4.000000                  NaN                    NaN   
50%          5.000000                  NaN                    NaN   
75%          5.000000                  NaN                    NaN   
max          5.000000                  NaN                    NaN   

       review_creation_date review_answer_timestamp  
count                100000                  100000  
unique                  637                   99010  
top     2017-12-19 00:00:00     2017-06-15 23:21:05  
freq                    466                       4  
first   2016-10-02 00:00:00     2016-10-07 18:32:28  
last    2018-08-31 00:00:00     2018-10-29 12:27:35  
mean                    NaN                     NaN  
std                     NaN                     NaN  
min                     NaN                     NaN  
25%                     NaN                     NaN  
50%                     NaN                     NaN  
75%                     NaN                     NaN  
max                     NaN                     NaN

Des valeurs manquantes dans le titre et les commentaires.  
Deux attributs qui sont des dates.  
Une évaluation peut concerner plusieurs commandes. 
Une commande peut avoir plusieurs évaluations à des dates différentes (réponses ou réactions aux réponses).  

In [10]:
# import du data set orders et premières informations

orders = pd.read_csv("archive/olist_orders_dataset.csv", 
                     parse_dates=["order_purchase_timestamp", "order_approved_at",
                                  "order_delivered_carrier_date", "order_delivered_customer_date",
                                  "order_estimated_delivery_date"])
print(orders.shape)
orders.head()

(99441, 8)


order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0                    2017-10-18  
1                    2018-08-13  
2                    2018-09-04  
3                    2017-12-15  
4                    2018-02-26

In [11]:
# statistiques sur orders

print(orders.dtypes)
orders.describe(include="all")

order_id                                 object
customer_id                              object
order_status                             object
order_purchase_timestamp         datetime64[ns]
order_approved_at                datetime64[ns]
order_delivered_carrier_date     datetime64[ns]
order_delivered_customer_date    datetime64[ns]
order_estimated_delivery_date    datetime64[ns]
dtype: object


<ipython-input-11-48bbb8b7e4cd>:4: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  orders.describe(include="all")
<ipython-input-11-48bbb8b7e4cd>:4: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  orders.describe(include="all")
<ipython-input-11-48bbb8b7e4cd>:4: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  orders.describe(include="all")
<ipython-input-11-48bbb8b7e4cd>:4: FutureWarning: Treat

order_id                       customer_id  \
count                              99441                             99441   
unique                             99441                             99441   
top     bff76a871750bd5f03d22c007cd7d26d  54acd3c8fe76a1805d6fe6d9a326f203   
freq                                   1                                 1   
first                                NaN                               NaN   
last                                 NaN                               NaN   

       order_status order_purchase_timestamp    order_approved_at  \
count         99441                    99441                99281   
unique            8                    98875                90733   
top       delivered      2018-03-31 15:08:21  2018-02-27 04:31:10   
freq          96478                        3                    9   
first           NaN      2016-09-04 21:15:19  2016-09-15 12:16:38   
last            NaN      2018-10-17 17:30:18  2018-09-03 17:40:06   

       order_delivered_carrier_date order_delivered_customer_date  \
count                         97658                         96476   
unique                        81018                         95664   
top             2018-05-09 15:48:00           2018-05-14 20:02:44   
freq                             47                             3   
first           2016-10-08 10:34:01           2016-10-11 13:46:32   
last            2018-09-11 19:48:28           2018-10-17 13:22:46   

       order_estimated_delivery_date  
count                          99441  
unique                           459  
top              2017-12-20 00:00:00  
freq                             522  
first            2016-09-30 00:00:00  
last             2018-11-12 00:00:00

In [12]:
orders["order_status"].unique()

array(['delivered', 'invoiced', 'shipped', 'processing', 'unavailable',
       'canceled', 'created', 'approved'], dtype=object)

5 attributs qui sont des dates.  
Quelques valeurs manquantes sur celles-ci, qui correspondent aux commandes qui n'ont pas abouti (order_status autre que delivered).  
96 478 commandes livrées.


In [13]:
# import du data set products

products = pd.read_csv("archive/olist_products_dataset.csv")
print(products.shape)
products.head()

(32951, 9)


product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

In [14]:
# statistiques sur products

print(products.dtypes)
products.describe(include="all")

product_id                     object
product_category_name          object
product_name_lenght           float64
product_description_lenght    float64
product_photos_qty            float64
product_weight_g              float64
product_length_cm             float64
product_height_cm             float64
product_width_cm              float64
dtype: object


product_id product_category_name  \
count                              32951                 32341   
unique                             32951                    73   
top     730e0343a68b7cebc35ad4d14cf9f7af       cama_mesa_banho   
freq                                   1                  3029   
mean                                 NaN                   NaN   
std                                  NaN                   NaN   
min                                  NaN                   NaN   
25%                                  NaN                   NaN   
50%                                  NaN                   NaN   
75%                                  NaN                   NaN   
max                                  NaN                   NaN   

        product_name_lenght  product_description_lenght  product_photos_qty  \
count          32341.000000                32341.000000        32341.000000   
unique                  NaN                         NaN                 NaN   
top                     NaN                         NaN                 NaN   
freq                    NaN                         NaN                 NaN   
mean              48.476949                  771.495285            2.188986   
std               10.245741                  635.115225            1.736766   
min                5.000000                    4.000000            1.000000   
25%               42.000000                  339.000000            1.000000   
50%               51.000000                  595.000000            1.000000   
75%               57.000000                  972.000000            3.000000   
max               76.000000                 3992.000000           20.000000   

        product_weight_g  product_length_cm  product_height_cm  \
count       32949.000000       32949.000000       32949.000000   
unique               NaN                NaN                NaN   
top                  NaN                NaN                NaN   
freq                 NaN                NaN                NaN   
mean         2276.472488          30.815078          16.937661   
std          4282.038731          16.914458          13.637554   
min             0.000000           7.000000           2.000000   
25%           300.000000          18.000000           8.000000   
50%           700.000000          25.000000          13.000000   
75%          1900.000000          38.000000          21.000000   
max         40425.000000         105.000000         105.000000   

        product_width_cm  
count       32949.000000  
unique               NaN  
top                  NaN  
freq                 NaN  
mean           23.196728  
std            12.079047  
min             6.000000  
25%            15.000000  
50%            20.000000  
75%            30.000000  
max           118.000000

In [15]:
products[products["product_category_name"].isnull()]

product_id product_category_name  \
105    a41e356c76fab66334f36de622ecbd3a                   NaN   
128    d8dee61c2034d6d075997acef1870e9b                   NaN   
145    56139431d72cd51f19eb9f7dae4d1617                   NaN   
154    46b48281eb6d663ced748f324108c733                   NaN   
197    5fb61f482620cb672f5e586bb132eae9                   NaN   
...                                 ...                   ...   
32515  b0a0c5dd78e644373b199380612c350a                   NaN   
32589  10dbe0fbaa2c505123c17fdc34a63c56                   NaN   
32616  bd2ada37b58ae94cc838b9c0569fecd8                   NaN   
32772  fa51e914046aab32764c41356b9d4ea4                   NaN   
32852  c4ceee876c82b8328e9c293fa0e1989b                   NaN   

       product_name_lenght  product_description_lenght  product_photos_qty  \
105                    NaN                         NaN                 NaN   
128                    NaN                         NaN                 NaN   
145                    NaN                         NaN                 NaN   
154                    NaN                         NaN                 NaN   
197                    NaN                         NaN                 NaN   
...                    ...                         ...                 ...   
32515                  NaN                         NaN                 NaN   
32589                  NaN                         NaN                 NaN   
32616                  NaN                         NaN                 NaN   
32772                  NaN                         NaN                 NaN   
32852                  NaN                         NaN                 NaN   

       product_weight_g  product_length_cm  product_height_cm  \
105               650.0               17.0               14.0   
128               300.0               16.0                7.0   
145               200.0               20.0               20.0   
154             18500.0               41.0               30.0   
197               300.0               35.0                7.0   
...                 ...                ...                ...   
32515            1800.0               30.0               20.0   
32589             800.0               30.0               10.0   
32616             200.0               21.0                8.0   
32772            1300.0               45.0               16.0   
32852             700.0               28.0                3.0   

       product_width_cm  
105                12.0  
128                20.0  
145                20.0  
154                41.0  
197                12.0  
...                 ...  
32515              70.0  
32589              23.0  
32616              16.0  
32772              45.0  
32852              43.0  

[610 rows x 9 columns]

32 951 produits répartis dans 73 catégories.  
610 produits qui ne sont pas catégorisés.

In [16]:
# import du data set categories

categories = pd.read_csv("archive/product_category_name_translation.csv")
print(categories.shape)
categories.head()

(71, 2)


product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor

In [17]:
# statistiques sur categories

print(categories.dtypes)
categories.describe(include="all")

product_category_name            object
product_category_name_english    object
dtype: object


product_category_name product_category_name_english
count                     71                            71
unique                    71                            71
top               cool_stuff                    cool_stuff
freq                       1                             1

Traduction de 71 (et pas 73 !!!) catégories du portugais à l'anglais.



# Creation des data sets à clusteriser

In [18]:
# création d'un DataFrame contenant customer_id, customer_unique_id et city_rank
# city_rank correspond au rang de la ville :
# 5 pour plus de 2 000 000, 4 pour plus de 1 000 000, 3 pour plus de 500 000,
# 2 pour plus de 200 000, 1 pour plus de 100 000 et 0 en dessous 
# (source https://en.wikipedia.org/wiki/List_of_cities_in_Brazil_by_population)

cities = pd.read_csv("villes.csv", header=None, names=["customer_city", "city_rank"])
cities["customer_city"] = cities["customer_city"].map(lambda x: x.lower())
cities["customer_city"] = cities["customer_city"].map(lambda x: x.strip(' '))
customers = raw_customers[["customer_id", "customer_unique_id", "customer_city"]]
customers = pd.merge(customers, cities, on="customer_city", how="left")
customers.drop(columns=["customer_city"], inplace=True)
customers = customers.fillna(0)
customers

customer_id                customer_unique_id  \
0      06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1      18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2      4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3      b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4      4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                 ...                               ...   
99436  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
99437  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
99438  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
99439  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
99440  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

       city_rank  
0            2.0  
1            3.0  
2            5.0  
3            2.0  
4            4.0  
...          ...  
99436        5.0  
99437        2.0  
99438        5.0  
99439        2.0  
99440        2.0  

[99441 rows x 3 columns]

In [19]:
# création d'un DataFrame contenant order_purchase_timestamp, order_id, customer_id, city_rank
# et customer_unique_id avec les commandes dont le order_status est delivered.
# Ce DataFrame, df_orders, est indéxé par date

df_orders = orders.loc[orders["order_status"]=="delivered",["order_id","customer_id","order_purchase_timestamp"]]
df_orders = pd.merge(df_orders, customers, on="customer_id", how="left")
df_orders = df_orders.set_index("order_purchase_timestamp")
df_orders.head()

order_id  \
order_purchase_timestamp                                     
2017-10-02 10:56:33       e481f51cbdc54678b7cc49136f2d6af7   
2018-07-24 20:41:37       53cdb2fc8bc7dce0b6741e2150273451   
2018-08-08 08:38:49       47770eb9100c2d0c44946d9cf07ec65d   
2017-11-18 19:28:06       949d5b44dbf5de918fe9c16f97b45f8a   
2018-02-13 21:18:39       ad21c59c0840e6cb83a9ceb5573f8159   

                                               customer_id  \
order_purchase_timestamp                                     
2017-10-02 10:56:33       9ef432eb6251297304e76186b10a928d   
2018-07-24 20:41:37       b0830fb4747a6c6d20dea0b8c802d7ef   
2018-08-08 08:38:49       41ce2a54c0b03bf3443c3d931a367089   
2017-11-18 19:28:06       f88197465ea7920adcdbec7375364d82   
2018-02-13 21:18:39       8ab97904e6daea8866dbdbc4fb7aad2c   

                                        customer_unique_id  city_rank  
order_purchase_timestamp                                               
2017-10-02 10:56:33       7c396fd4830fd04220f754e42b4e5bff        5.0  
2018-07-24 20:41:37       af07308b275d755c9edb36a90c618231        1.0  
2018-08-08 08:38:49       3a653a41f6f9fc3d2a113cf8398680e8        0.0  
2017-11-18 19:28:06       7c142cf63193a1473d2e66489a9ae977        1.0  
2018-02-13 21:18:39       72632f0f9dd73dfee390c9b22eb56dd6        3.0

In [20]:
# déclinaisons de df_orders par périodes 

df_orders_quarter = df_orders.to_period(freq="Q")
df_orders_month = df_orders.to_period(freq="M")
df_orders_week = df_orders.to_period(freq="W")


In [21]:
# stockage des DataFrames par période dans des dictionnaires

list_quarters = df_orders_quarter.sort_index(inplace=False).index.unique()
dict_df_orders_quarter = dict()
for index in list_quarters:
    df = df_orders_quarter.loc[index]
    if isinstance(df, pd.core.series.Series):
        df = df.to_frame().transpose()
    dict_df_orders_quarter[str(index)] = df

list_month = df_orders_month.sort_index(inplace=False).index.unique()
dict_df_orders_month = dict()
for index in list_month:
    df = df_orders_month.loc[index]
    if isinstance(df, pd.core.series.Series):
        df = df.to_frame().transpose()
    dict_df_orders_month[str(index)] = df
del dict_df_orders_month["2016-09"]

In [22]:
# regroupement des catégories

grouped_categories = {
"body" : ["beleza_saude", "perfumaria", "fraldas_higiene"],
"electronic_computers" : ["informatica_acessorios", "telefonia", "tablets_impressao_imagem", "telefonia_fixa",
                        "consoles_games", "eletronicos", "pcs", "pc_gamer"],
"cars" : ["automotivo"],
"house" : ["cama_mesa_banho", "moveis_decoracao", "utilidades_domesticas", "ferramentas_jardim", "eletroportateis",
         "climatizacao", "moveis_cozinha_area_de_servico_jantar_e_jardim", "moveis_escritorio", "eletrodomesticos", 
         "casa_conforto", "moveis_colchao_e_estofado", "moveis_sala", "moveis_quarto", "eletrodomesticos_2", 
         "la_cuisine", "casa_conforto_2", "portateis_casa_forno_e_cafe", 
         "portateis_cozinha_e_preparadores_de_alimentos"],
"leisure_gifts" : ["esporte_lazer", "relogios_presentes", "papelaria", "cool_stuff", "artigos_de_festas", 
                 "artigos_de_natal", "flores"],
"food" : ["alimentos_bebidas", "market_place", "alimentos", "bebidas"],
"family_pets" : ["bebes", "brinquedos", "pet_shop", "other"],
"clothes_fashion" : ["fashion_bolsas_e_acessorios", "malas_acessorios", "fashion_roupa_masculina", 
                   "fashion_underwear_e_moda_praia", "fashion_esporte", "fashion_roupa_feminina", 
                   "fashion_roupa_infanto_juvenil", "fashion_calcados"],
"art_music_books" : ["audio", "livros_tecnicos", "instrumentos_musicais", "artes", "livros_interesse_geral", 
                   "livros_importados", "cine_foto", "musica", "cds_dvds_musicais", "dvds_blu_ray", 
                   "artes_e_artesanato"],
"professionnal" : ["construcao_ferramentas_construcao", "agro_industria_e_comercio", "construcao_ferramentas_jardim", 
                 "construcao_ferramentas_ferramentas", "casa_construcao", "construcao_ferramentas_iluminacao",
                 "industria_comercio_e_negocios", "construcao_ferramentas_seguranca", "sinalizacao_e_seguranca",
                 "seguros_e_servicos"]
}

In [23]:
# fonction de remplissage des features

# liste des moyens de paiement
list_payment_types = list(payments["payment_type"].unique())

# liste des catégories
list_categories = list(products["product_category_name"].dropna().unique())
list_categories.append("other")

def features(data):
    """
    Cette fonction reçoit en paramètre un DataFrame du type df_orders et renvoie
    un DataFrame avec les customer_unique_id en index, complèté avec les features suivantes :
        - nombre de commandes ("orders_count")
        - taux moyen de satisfaction ("review_score_mean")
        - utilisateur ou non des commentaires ("review_user")
        - nombre d'articles commandés ("nb_items")
        - somme totale dépensée ("total spent")
        - somme moyenne dépensée par article ("mean_spent")
        - nombre d'articles commandés par catégorie (cf list_categories)
        - nombre d'utilisation des moyens de paiements (cf list_payment type)
    """
    # on ne garde qu'une valeur de chaque customer_unique_id avec son city_rank
    returned_data = data.groupby("customer_unique_id").city_rank.mean()
    n_obs = returned_data.shape[0]
    
    # on ajoute le nombre de commandes
    df = data.groupby("customer_unique_id").agg(
        orders_count = pd.NamedAgg("order_id", "count"))
    returned_data = pd.merge(returned_data, df, on="customer_unique_id")
    
    # on ajoute le taux moyen de satisfaction et l'usage des commentaires
    df = pd.merge(data, reviews[["order_id", "review_score", "review_comment_title", "review_comment_message"]],
                  on="order_id", how="left")
    df1 = df.groupby("customer_unique_id").agg(
        review_score_mean = pd.NamedAgg("review_score", "mean"))
    returned_data = pd.merge(returned_data, df1, on="customer_unique_id")
    df["review_user"] =  ((~(df["review_comment_title"].isnull())) | (~(df["review_comment_message"].isnull())))
    df1 = df.groupby("customer_unique_id").agg(
        review_user = pd.NamedAgg("review_user", "sum"))
    returned_data = pd.merge(returned_data, df1, on="customer_unique_id")
    
    # on ajoute le nombre d'objets, leurs catégories, la somme totale et la dépense moyenne
    df = pd.merge(data, items[["order_id", "order_item_id", "product_id", "price", "freight_value"]],
                  on="order_id", how="left")
    df = pd.merge(df, products[["product_id", "product_category_name"]], 
                  on="product_id", how="left")
    df["product_category_name"] = df["product_category_name"].fillna("other")
    df1 = df.groupby("customer_unique_id").agg(
        nb_items = pd.NamedAgg("product_id", "count"),
        total_prices = pd.NamedAgg("price", "sum"),
        total_freight = pd.NamedAgg("freight_value", "sum"))
    returned_data = pd.merge(returned_data, df1, on="customer_unique_id")
    returned_data = returned_data.assign(total_spent=returned_data["total_prices"]+returned_data["total_freight"])
    returned_data.drop(columns=["total_prices", "total_freight"], inplace=True)
    returned_data = returned_data.assign(mean_spent=returned_data["total_spent"]/returned_data["nb_items"])
    df1 = df.groupby("customer_unique_id").product_category_name.value_counts()
    df1 = df1.unstack(level=-1, fill_value=0)
    returned_data = pd.merge(returned_data, df1, on="customer_unique_id")
    missing_cols = set(list_categories) - set(df1.columns)
    for cat_name in missing_cols:
        returned_data = returned_data.assign(new_col=np.zeros(n_obs, dtype=int))
        returned_data.rename(columns={"new_col" : cat_name}, inplace=True)
    
    # on regroupe les catégories
    for key_list, list_cat in grouped_categories.items():
        returned_data[key_list] = returned_data[list_cat].sum(axis=1)
        returned_data.drop(columns=list_cat, inplace=True)
    
    # on ajoute les moyens de paiement utilisés
    df = pd.merge(data, payments[["order_id", "payment_type"]], on="order_id", how="left")
    df = df.groupby("customer_unique_id").payment_type.value_counts()
    df = df.unstack(level=-1, fill_value=0)
    returned_data = pd.merge(returned_data, df, on="customer_unique_id")
    missing_cols = set(list_payment_types) - set(df.columns)
    for payment_type in missing_cols:
        returned_data = returned_data.assign(new_col=np.zeros(n_obs, dtype=int))
        returned_data.rename(columns={"new_col" : payment_type}, inplace=True)
     
    return returned_data

In [24]:
# remplissage des features pour chaque période et stockage dans des dictionnaires

dict_quarter = dict()
for period, df in dict_df_orders_quarter.items():
    try :
        dict_quarter[period] = features(df)
    except :
        print(f"Période {period} inexploitable")
    
dict_month = dict()
for period, df in dict_df_orders_month.items():
    try :
        dict_month[period] = features(df)
    except :
        print(f"Période {period} inexploitable")

Période 2016Q3 inexploitable
Période 2016-12 inexploitable


In [25]:
# enregistrement des DataFrames utilisables

pickle.dump(dict_quarter, open("pickle_dict_quarter.pkl", "wb"))
pickle.dump(dict_month, open("pickle_dict_month.pkl", "wb"))

In [26]:
# enregistrement des catégories et de la liste des type de paiement

pickle.dump(grouped_categories, open("pickle_grouped_categories.pkl", "wb"))
pickle.dump(list_payment_types, open("pickle_list_payment_types.pkl", "wb"))